contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -92,45 +92,90 @@ contract MixinExchangeWrapper is
         bytes[] memory signatures
     )
         internal
-        returns (FillResults memory totalFillResults)
+        returns (
+            FillResults memory totalFillResults,
+            uint256 wethSpentAmount,
+            uint256 makerAssetAcquiredAmount
+        )
     {
-        bytes memory makerAssetData = orders[0].makerAssetData;
-        bytes memory wethAssetData = WETH_ASSET_DATA;
-
         uint256 ordersLength = orders.length;
-        for (uint256 i = 0; i != ordersLength; i++) {
-
-            // We assume that asset being bought by taker is the same for each order.
-            // We assume that asset being sold by taker is WETH for each order.
-            orders[i].makerAssetData = makerAssetData;
-            orders[i].takerAssetData = wethAssetData;
 
-            // Calculate the remaining amount of WETH to sell
-            uint256 remainingTakerAssetFillAmount = _safeSub(wethSellAmount, totalFillResults.takerAssetFilledAmount);
+        for (uint256 i = 0; i != ordersLength; i++) {
+            require(
+                orders[i].makerAssetData.equals(orders[0].makerAssetData),
+                "MAKER_ASSET_MISMATCH"
+            );
 
-            // Attempt to sell the remaining amount of WETH
-            FillResults memory singleFillResults = _fillOrderNoThrow(
-                orders[i],
-                remainingTakerAssetFillAmount,
-                signatures[i]
+            // The remaining amount of WETH to sell
+            uint256 remainingTakerAssetFillAmount = _safeSub(
+                wethSellAmount,
+                wethSpentAmount
             );
+            FillResults memory singleFillResults;
+
+            // Percentage fee
+            if (orders[i].makerAssetData.equals(orders[i].takerFeeAssetData)) {
```

Comment:
Shouldnt this check that the order's `takerAssetData` matches the `takerFeeAssetData`, since we don't want to over-sell the `takerAsset`?

Reply:
I see, this is yet another case. Let's write out all of the different cases and see if we can simplify the logic at all.